1.2-Introduction-to-Musig.ipynb

```diff
@@ -0,0 +1,186 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 1: Public Key Generation\n",
+    "\n",
+    "* Due to linear nature of Schnorr, public keys, nonces and signatures can be aggregated.\n",
+    "* Due to the [Key cancellation](https://tlu.tarilabs.com/cryptography/digital_signatures/introduction_schnorr_signatures.html#key-cancellation-attack) attack, signers must first commit to individual public keys (Challenge factor).\n",
+    "* Individual key pairs must be multiplied with respective challenge.\n",
+    "\n",
+    "\n",
+    "![test](images/musig_intro_1.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1.1 Programming Exercise: Compute 3-of-3 musig public key.\n",
+    "* Use `generate_musig_key([ECPubKey0, ECPubKey1, ...])`.\n",
+    "* returns `c_map, pk_musig`\n",
+    "* `c_map` contains `ECPubKey_i, challenge_data_i` key - value pairs.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Compute key pairs.\n",
+    "\n",
+    "\n",
+    "# Compute key Challenges.\n",
+    "\n",
+    "\n",
+    "# Multiply key pairs by challenge factor.\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 2: Nonce Generation\n",
+    "\n",
+    "* Nonces can be aggregated like public keys and signatures.\n",
+    "* Security [Proof](https://eprint.iacr.org/2018/068.pdf) requires nonces are randomly generated independently of each other.\n",
+    "* An initial nonce commitment round is necessary to ensure no nonce is derived as a function of another.\n",
+    "\n",
+    "\n",
+    "![test](images/musig_intro_2.jpg)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.1 Programming Exercise: Compute 3-of-3 musig nonce.\n",
+    "\n",
+    "* `generate_schnorr_nonce()` returns a ECkey with with a jacobi symbol of 1.\n",
+    "* `aggregate_schnorr_nonces()` returns a ECPubKey and a boolean (indicates nonce negation).\n",
+    "* *Negation of nonce `k_i` isn't necessary, since it is performed by `sign_musig` in next exercise.*"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate nonces.\n",
+    "\n",
+    "\n",
+    "# Skip nonce point commitment round (we assume nonce commitments are correct).\n",
+    "\n",
+    "\n",
+    "# Aggregate nonces.\n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 3: Signature Generation\n",
+    "\n",
+    "* Signatures are generated individually.\n",
+    "\n",
+    "![test](images/musig_intro_3.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.1 Programming exercise: Compute aggregated musig signature.\n",
```

Comment:
It would be good if you could show _why_ the aggregate signature is valid for the aggregated pubkey